tools/cache_creator/cache_info.cpp

```diff
@@ -0,0 +1,270 @@
+                             entryIdx, currEntryOffset);
+
+    currData += EntrySize;
+    currEntryInfo.entryBlob = {currData, currEntryBlobSize};
+    currData += currEntryBlobSize;
+
+    llvm::MD5 md5;
+    md5.update(currEntryInfo.entryBlob);
+    llvm::MD5::MD5Result result = {};
+    md5.final(result);
+    currEntryInfo.entryMD5Sum = result.digest();
+
+    if (sourceElfMD5ToIdentifier) {
+      auto sourceElfIt = sourceElfMD5ToIdentifier->find(currEntryInfo.entryMD5Sum);
+      if (sourceElfIt != sourceElfMD5ToIdentifier->end())
+        currEntryInfo.sourceElfBlobIdentifier = sourceElfIt->second;
+    }
+
+    currEntryInfo.sourceElfBlobIdentifier = llvm::None;
+
+    entriesInfoOut.emplace_back(std::move(currEntryInfo));
+    ++entryIdx;
+  }
+
+  return llvm::Error::success();
+}
+
+llvm::raw_ostream &operator<<(llvm::raw_ostream &os, const PublicVkHeaderInfo &info) {
+  assert(info.publicHeader);
+  const vk::PipelineCacheHeaderData &header = *info.publicHeader;
+
+  os << "=== Vulkan Pipeline Cache Header ===\n"
+     << "header length:\t\t" << header.headerLength << "\n"
+     << "header version:\t\t" << header.headerVersion << "\n"
+     << "vendor ID:\t\t" << llvm::format("0x%" PRIx32, header.vendorID) << "\n"
+     << "device ID:\t\t" << llvm::format("0x%" PRIx32, header.deviceID) << "\n"
+     << "pipeline cache UUID:\t" << cc::uuidToHexString(header.UUID) << "\n";
+  return os;
+}
+
+llvm::raw_ostream &operator<<(llvm::raw_ostream &os, const BinaryCachePrivateHeaderInfo &info) {
+  assert(info.privateHeader);
+  const vk::PipelineBinaryCachePrivateHeader &header = *info.privateHeader;
+
+  os << "=== Pipeline Binary Cache Private Header ===\n"
+     << "header length:\t" << sizeof(header) << "\n"
+     << "hash ID:\t" << llvm::format_bytes(header.hashId, llvm::None, sizeof(header.hashId)) << "\n"
+     << "content size:\t" << info.contentBlobSize << "\n";
+  return os;
+}
+
+llvm::raw_ostream &operator<<(llvm::raw_ostream &os, const BinaryCacheEntryInfo &info) {
+  assert(info.entryHeader);
+  const vk::BinaryCacheEntry &header = *info.entryHeader;
+
+  os << "\t*** Entry " << info.idx << " ***\n"
+     << "\thash ID:\t\t" << llvm::format_bytes(header.hashId.bytes, llvm::None, 16, 1) << "\n"
+     << "\tdata size:\t\t" << header.dataSize << "\n"
+     << "\tcalculated MD5 sum:\t" << info.entryMD5Sum << "\n"
+     << "\tmatched source blob:\t" << info.sourceElfBlobIdentifier.getValueOr("<none>") << "\n";
+  return os;
+}
+
+// =====================================================================================================================
+// Recursively finds all ELF files in the directory and creates a map from ELF MD5 sums to their paths on disk.
+// This only considers regular files with the `.elf` extension.
+//
+// @param dir : The directory to search
+// @returns : Map from ELF MD5 sums to their paths on disk
+llvm::StringMap<std::string> collectSourceElfMD5Sums(llvm::Twine dir) {
```

Comment:
The function name can be a little confusing.  What is a Source Elf? You are potentially adding extra context to the name that is not needed for this function.  Also "collect" is a generic and does not reflect that a map is being created. I would expect a function with this name to return a set of MD5 sums.  Something like `mapMD5SumsToElfFilePath` might be a better name.

Also, what will happen if two files have the same MD5 sum?  That should be mentioned in the comments for the function.

Reply:
I struggled to find a good name for this function, I like your suggestion, it's more descriptive.